rectools/models/sasrec.py

```diff
@@ -922,18 +928,54 @@ def on_train_start(self) -> None:
 
     def training_step(self, batch: Dict[str, torch.Tensor]) -> torch.Tensor:
         """TODO"""
+        x, y, w = batch["x"], batch["y"], batch["yw"]
         if self.loss == "softmax":
-            return self._calc_softmax_loss(batch)
+            logits = self._get_logits(x)
+            return self._calc_softmax_loss(logits, y, w)
         if self.loss == "BCE":
-            return self._calc_bce_loss(batch)
+            negatives = batch["negatives"]
+            logits = self._get_pos_neg_logits(x, y, negatives)
+            return self._calc_bce_loss(logits, y, w)
         if self.loss == "gBCE":
-            return self._calc_gbce_loss(batch)
+            negatives = batch["negatives"]
+            logits = self._get_pos_neg_logits(x, y, negatives)
+            return self._calc_gbce_loss(logits, y, w, negatives)
 
         raise ValueError(f"loss {self.loss} is not supported")
 
-    def _calc_softmax_loss(self, batch: Dict[str, torch.Tensor]) -> torch.Tensor:
-        x, y, w = batch["x"], batch["y"], batch["yw"]
-        item_embs, session_embs = self.forward(x)
+    def _get_logits(self, x: torch.Tensor) -> torch.Tensor:
+        item_embs, session_embs = self.torch_model(x)
+        logits = session_embs @ item_embs.T
+        return logits
+
+    def _get_pos_neg_logits(self, x: torch.Tensor, y: torch.Tensor, negatives: torch.Tensor) -> torch.Tensor:
+        # [n_items + n_item_extra_tokens, n_factors], [batch_size, session_max_len, n_factors]
+        item_embs, session_embs = self.torch_model(x)
+        pos_neg_embs = torch.cat([y.unsqueeze(-1), negatives], dim=-1)  # [batch_size, session_max_len, n_negatives + 1]
```

Comment:
```suggestion
        pos_neg = torch.cat([y.unsqueeze(-1), negatives], dim=-1)  # [batch_size, session_max_len, n_negatives + 1]
```
That's not `embs` because it's not embeddings, just ids.